src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -629,6 +634,98 @@ theorem shiftLeft_shiftLeft {w : Nat} (x : BitVec w) (n m : Nat) :
     getLsb (x >>> i) j = getLsb x (i+j) := by
   unfold getLsb ; simp
 
+/-! ### sshiftRight -/
+
+theorem sshiftRight_eq {x : BitVec n} {i : Nat} :
+    x.sshiftRight i = BitVec.ofInt n (x.toInt >>> i) := by
+  apply BitVec.eq_of_toInt_eq
+  simp [BitVec.sshiftRight]
+
+/-- The MSB of a bitvector is `false` iff its value is smaller than 2^(half the bitwidth). -/
+theorem msb_eq_false_iff_small {x : BitVec w} :
+    x.msb = false ↔ 2 * x.toNat < 2^w  := by
+  cases w <;> simp [Nat.pow_succ, Nat.mul_comm _ 2, msb_eq_decide]
+
+/-- The MSB of a bitvector is `true` iff its value interpreted as a natural number in 2s complement
+  is larger than 2^(half the bitwidth). -/
+theorem msb_eq_true_iff_large {x : BitVec w} :
+    x.msb = true ↔ 2 * x.toNat ≥ 2^w  := by
+  simp [show x.msb = true ↔ ¬(x.msb = false) by simp, msb_eq_false_iff_small]
+
+/-- Equation theorem for 'Int.sub' when both arguments are 'negSucc' -/
+theorem Int.toNat_sub_toNat_eq_negSucc_ofLt {n m : Nat} (hlt : n < m) :
+  (n : Int) - (m : Int) = (Int.negSucc (m - 1 - n)) := by
```

Comment:
I am very skeptical of introducing `Int.negSucc` into the proof at all. Can you explain why? Even better, can you avoid it?

Reply:
Also, these lemmas should be further down, immediately before the theorem that uses them.

Reply:
e.g. I would try to keep negSucc out of it, but wrap the fact critical fact `Int.shiftRight_negSucc` into a `Int.shiftRight_sub_of_lt`.

Reply:
@semorrison that's fair enough, I'm happy to refactor it. 

when I write `Int.shiftRight_sub_of_lt` in `Init/Data/Int/Bitwise.lean`, I imagine I would still need the (private) theorem `Int.toNat_sub_toNat_eq_negSucc_ofLt `. Is there a different way to arrive at the same result that I'm missing? If not, I'd prefer it lives somewhere that helps the next person who stumbles upon the same need.

Reply:
@semorrison actually, on second thought, I prefer the current implementation: The lemma name `Int.shiftRight_sub_of_lt` does not convey the full extent of the theorem statement, doesn't it? The equivalent `sshiftRight` theorem would be: 

```lean
diff --git a/src/Init/Data/Int/Bitwise/Lemmas.lean b/src/Init/Data/Int/Bitwise/Lemmas.lean
index c4b29d23ce..01ade45c2e 100644
--- a/src/Init/Data/Int/Bitwise/Lemmas.lean
+++ b/src/Init/Data/Int/Bitwise/Lemmas.lean
@@ -19,7 +19,21 @@ theorem natCast_shiftRight (n s : Nat) : ((↑n) : Int) >>> s = n >>> s := rfl
 theorem shiftRight_negSucc (m n : Nat) :
     -[m+1] >>> n = -[m >>>n +1] := rfl

+/-- Equation theorem for 'Int.sub' when the arguments are `.ofNat` and
+  the result is known to be negative. -/
+private theorem toNat_sub_toNat_eq_negSucc_ofLt {n m : Nat} (hlt : n < m) :
+  (n : Int) - (m : Int) = (Int.negSucc (m - 1 - n)) := by
+  rw [Int.negSucc_eq] -- TODO: consider adding this to omega cleanup set.
+  omega
+
+/-- shiftRight of a subtraction is evaluated when
+  the result of the subtraction is negative. -/
+theorem toNat_sub_toNat_shiftRight_eq_ofLt {m n i: Nat} (h : m < n) :
+  ((m : Int) - (n : Int)) >>> i = -[((n - 1 - m) >>> i) +1] := by
+    rw [toNat_sub_toNat_eq_negSucc_ofLt (by omega)]
+    rw [shiftRight_negSucc]
```

but this theorem seems *far too specialized* for my taste --- I much prefer the status quo, where the proof factors as two steps: first evaluate the subtraction, then evaluate the shift, knowing that the subtraction was a negative number. 

Maybe I misunderstood the refactor you were suggesting?

Reply:
Okay, well, I haven't tried to understand the rest of your proof yet so I may be wrong here. I had in mind a theorem that didn't even have an `Int.negSucc` on the RHS. That seems really smelly --- `Int.negSucc` is necessary when setting up the basics, but it is an implementation detail and should never appear in arithmetic lemmas about `Int`.

Reply:
Hopefully, here's a diff that's more tasteful:

```diff
 /-- If the msb is `true`, the arithmetic shift right equals negating, then logical shifting right, then negating again.
   The double negation preserves the lower bits that have been shifted,
   and the outer negation ensures that the high bits are '1'. -/
@@ -693,16 +694,17 @@ theorem sshiftRight_eq_of_msb_true {x : BitVec w} {s : Nat} (h : x.msb = true) :
     have hxbound : (2 * x.toNat ≥ 2 ^ (w + 1)) := BitVec.msb_eq_true_iff_two_mul_ge.mp h
     have hxbound' : ¬ (2 * x.toNat < 2 ^ (w + 1)) := by omega
     simp only [hxbound', ↓reduceIte, toNat_ofInt, toNat_not, toNat_ushiftRight]
-    rw [Int.toNat_sub_toNat_eq_negSucc_ofLt (by omega)]
-    rw [Int.shiftRight_negSucc]
+    rw [← Int.subNatNat_eq_coe]
+    rw [Int.subNatNat_of_lt (by omega)]
+    simp
     rw [Int.mod_negSucc_eq]
-    simp only [Int.natAbs_ofNat, Nat.succ_eq_add_one]
+    simp [Int.natAbs_ofNat, Nat.succ_eq_add_one]
     rw [Int.subNatNat_of_le (by omega)]
     simp only [Int.toNat_ofNat]
     rw [Nat.mod_eq_of_lt]
-    · omega
+    · rw [Nat.sub_right_comm]
+      omega
     · rw [Nat.shiftRight_eq_div_pow]
-      suffices (2 ^ (w + 1) - 1 - x.toNat) / 2 ^ s < 2 ^ (w + 1) - 1 by omega
       apply Nat.lt_of_le_of_lt (Nat.div_le_self _ _) (by omega)
```

I switched to using `subNatNat` to rewrite the statement as I wanted. It feels obscure to me, but I guess that's just knowing the API :)

Reply:
Does look better to me. Sorry this was only a superficial comment. I'll try to look closely later tonight.